### PR TITLE
python/high scores Mentor Notes

### PR DESCRIPTION
**Pull request by BethanyG**
First shot at mentor notes for High Scores.  Needs some go-overs to catch anything I've missed.

### CODE REVIEW

**Comment by SleeplessByte** [comment]
Nice job @BethanyG . This looks very elaborate. 

One quick skimming suggestion: you have a lot of whitespace between each reasonable solution. You might want to prefix these solutions with a line or two about what is used. E.g. "solution using classes, solution using minmax, solution using package".

File: tracks/python/exercises/high-scores/mentoring.md [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+-  `self` also denotes _**object attributes**_  :  `self.scores`, `self.sorted_scores`.  Data variables meant to be part of an object need `self` as part of their declaration.   Object attribute values can be changed _**per instance**_.  What are the potential downsides to this?
+- Code inside methods runs _**on request**_.  Object methods are accessed from _outside_ the class using "dot notation" and the _**object instance name**_  (e.g. `my_scores.personal_top() ` or `my_scores.latest()`).  Object methods are accessed from _inside_ the class via `self`:  `self.personal_top()`, `self.latest()`.  What are the upsides/downsides of having code run on request vs running on instantiation?
+- In Python 3.5 and above, classes can be declared without `(object)` because it's **inferred**.  The parenthesis & class name only need to be included when you are inheriting directly from a class that's _**not**_ `object`.  The boilerplate has the `(object)` syntax for backward compatibility.
+
+  
+### Resources
+
+#### links into Python 3 documentation
```
Comment (cmccandless) [inline tracks/python/exercises/high-scores/mentoring.md]: Headings should be Title-Cased.
```suggestion
#### Links to Python 3 Documentation
```

File: tracks/python/exercises/high-scores/mentoring.md [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+-  [**A First Look at Classes**](https://docs.python.org/3/tutorial/classes.html#a-first-look-at-classes) 
+-  [**Data Structures in Python**](https://docs.python.org/3/tutorial/datastructures.html).  Describes list object methods, uses for lists, and comprehensions.
+- [**`list.sort()`**](https://docs.python.org/3/library/stdtypes.html#list.sort) **|** [**`sorted()`**](https://docs.python.org/3/library/functions.html#sorted)
+-  [**Sorting HOW TO**](https://docs.python.org/3/howto/sorting.html)  Tips theories and "recipes" for using lists and sorting to manage data in Python.
+-  [**`max()` built-in function**](https://docs.python.org/3/library/functions.html#max) 
+- [**`heapq` & `heapq.nlargest`**](https://docs.python.org/3/library/heapq.html#heapq.nlargest).  A _heap_ implementation.  Can be used in place of `sorted()` or `max()` for small sets of values.
+
+#### blogs and tutorials
```
Comment (cmccandless) [inline tracks/python/exercises/high-scores/mentoring.md]: Headings should be Title-Cased.
```suggestion
#### Blogs and Tutorials
```

File: tracks/python/exercises/high-scores/mentoring.md [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+  
+### Talking points
+
+#### copying data and lists
+- Python is _pass by object reference_.  You cannot copy list data by assigning a new variable name, you must copy data by calling `.copy()` or using slice syntax `copied_data = data[:]`
+- **Indexing and slicing work from either direction**.  Positive numbers from the left (starting at 0), negative numbers from the right (starting at -1).
+- Why _wouldn't_ you use `heapq.nlargest` for getting the top three scores?  What other strategies could you use within the class to create and manage the "top three" scores?  How does your strategy change if there are incoming scores to manage on top of the existing scores?
+
+  
+#### classes
```
Comment (cmccandless) [inline tracks/python/exercises/high-scores/mentoring.md]:
```suggestion
#### Classes
```

File: tracks/python/exercises/high-scores/mentoring.md [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+-  Use built-in functions like **`max(self.scores)`** -OR- **`sorted(self.scores, reverse=True)[0]`**  over looping manually.   It's is more efficient than iterating through `self.scores` to find the largest value.
+-  Using **`sorted(self.scores, reverse=True)[:3]`**  is preferable to  iterating through scores to find the three largest values.
+- Checking  **`len(self.scores) < 3`**  is unnecessary if a list slice is used to make the three highest scores list.  Python will return a copy of the entire list if it is shorter than the requested list slice.
+- A _negative index_ (e.g. **`self.scores[-1]`**) is preferred over performing a length-check  to get the index of the last element (e.g.**`self.scores[len(self.scores)-1]`**). 
+- The syntax of  **`sorted(self.scores, reverse=True)[:3]`** is preferred over  **`sorted(self.scores, reverse=True)[0:3]`**.  All slices default to zero unless otherwise specified - so convention is to leave it off.
+
+  
+### Talking points
+
+#### copying data and lists
```
Comment (cmccandless) [inline tracks/python/exercises/high-scores/mentoring.md]:
```suggestion
#### Copying Data and Lists
```

File: tracks/python/exercises/high-scores/mentoring.md [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+  Solutions that use `sort()` without first _copying_ the data in `self.scores`  erase the "latest" score.  If `latest()` is called after the `sort()`, it will be inaccurate.
+-  Use built-in functions like **`max(self.scores)`** -OR- **`sorted(self.scores, reverse=True)[0]`**  over looping manually.   It's is more efficient than iterating through `self.scores` to find the largest value.
+-  Using **`sorted(self.scores, reverse=True)[:3]`**  is preferable to  iterating through scores to find the three largest values.
+- Checking  **`len(self.scores) < 3`**  is unnecessary if a list slice is used to make the three highest scores list.  Python will return a copy of the entire list if it is shorter than the requested list slice.
+- A _negative index_ (e.g. **`self.scores[-1]`**) is preferred over performing a length-check  to get the index of the last element (e.g.**`self.scores[len(self.scores)-1]`**). 
+- The syntax of  **`sorted(self.scores, reverse=True)[:3]`** is preferred over  **`sorted(self.scores, reverse=True)[0:3]`**.  All slices default to zero unless otherwise specified - so convention is to leave it off.
+
+  
+### Talking points
```
Comment (cmccandless) [inline tracks/python/exercises/high-scores/mentoring.md]:
```suggestion
### Talking Points
```

File: tracks/python/exercises/high-scores/mentoring.md [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+    def personal_best(self):
+        return self.sorted_scores[-1]
+
+    def personal_top_three(self):
+        return self.sorted_scores[-1:-4:-1]
+```
+
+
+
+### Common suggestions
```
Comment (cmccandless) [inline tracks/python/exercises/high-scores/mentoring.md]:
```suggestion
### Common Suggestions
```

File: tracks/python/exercises/high-scores/mentoring.md [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+
+In tracking scores, students also use:
+
+-  List indexing and slicing syntax
+-  Strategies for computing/returning the maximum values in a list
+-  Strategies for sorting/filtering lists 
+
+  
+
+### Reasonable solutions
```
Comment (cmccandless) [inline tracks/python/exercises/high-scores/mentoring.md]:
```suggestion
### Reasonable Solutions
```

File: tracks/python/exercises/high-scores/mentoring.md [hunk 8]
```diff
@@ -0,0 +1,113 @@
+### Problem summary and challenges
```
Comment (cmccandless) [inline tracks/python/exercises/high-scores/mentoring.md]:
```suggestion
### Problem Summary and Challenges
```

**Comment by BethanyG** [comment]
@cmccandless  _Thank You_.  Wasn't clear on which way was preferred.  At least I was consistently wrong.  ;-)  Will push one more change related to @SleeplessByte comment on the Reasonable Solutions shortly.

**Comment by SleeplessByte** [comment]
@BethanyG you're the best ❤️ . Thanks for the review @cmccandless 🔥

**Pull request merged**